src/main/scala/rise/core/types/Constraints.scala

```diff
@@ -44,9 +44,12 @@ case class NatCollectionConstraint(a: NatCollection, b: NatCollection)
 }
 
 object Constraint {
-  def solve(cs: Seq[Constraint], trace: Seq[Constraint])
+  val canBeSubstituted : Seq[Kind.Identifier] => Kind.Identifier with Kind.Explicitness => Boolean =
+    preserve => i => !(preserve.contains(i) || i.isExplicit)
```

Comment:
would it make sense to also collect explicit variabels into `preserve` to simplify this test?

Reply:
it would then make sense to replace the `asImplicit`/`asExplicit` kind of transformations during type inference into `preserve` additions and substractions?
https://github.com/rise-lang/shine/pull/135/files#diff-b111bbb431e2b7263963fdc5d76146aef69e6a754dc981feb1d410e3feb28605L140-L147

Reply:
I think it depends on what we do with explicitness information: if we distinguish them by having two nodes `BoundedVar` and `GloballyFreeVar` (along the lines of what @michel-steuwer was suggesting), then constraint solving would still have to deal with them. @Bastacyclop should we leave this as it is here, and I open an issue for how to encode explicitness metadata and I mention it there?

Reply:
Opened an issue #151